cpukit/score/src/schedulerstrongapa.c

```diff
@@ -3,339 +3,820 @@
-      &node->Base.Base.Node.Chain,
-      &node->Ready_queue,
-      &self->Bit_map
-    );
-  } else {
-    _Scheduler_priority_Ready_queue_enqueue_first(
-      &node->Base.Base.Node.Chain,
-      &node->Ready_queue,
-      &self->Bit_map
-    );
+  filter_cpu = _Thread_Get_CPU( filter->user );
+  CPU = self->CPU;
+  cpu_max = _SMP_Get_processor_maximum();
+
+  for ( cpu_index = 0 ; cpu_index < cpu_max ; ++cpu_index ) {
+    CPU[ cpu_index ].visited = false;
   }
+
+  rear = rear + 1;
+  CPU[ rear ].cpu = filter_cpu;
+  CPU[ _Per_CPU_Get_index( filter_cpu ) ].visited = true;
+
+  highest_ready = _Scheduler_strong_APA_Find_highest_ready(
+                    self,
+                    front,
+                    rear
+                  );
+
+  if ( highest_ready != filter ) {
+    /*
+     * Backtrack on the path from
+     * filter_cpu to highest_ready, shifting along every task.
+     */
+
+    node = _Scheduler_strong_APA_Node_downcast( highest_ready );
+    /*
+     * Highest ready is not just directly reachable from the victim cpu
+     * So there is need for task shifting .
+     */
+    while ( node->cpu_to_preempt !=  filter_cpu ) {
+      curr_node = &node->Base.Base;
+      next_node = 
+          _Scheduler_strong_APA_Get_scheduled( self, node->cpu_to_preempt );
+
+      _Scheduler_SMP_Preempt(
+        context,
+        curr_node,
+        next_node,
+        _Scheduler_strong_APA_Allocate_processor
+      );
+
+      if( curr_node == highest_ready ) {
+        _Scheduler_strong_APA_Move_from_ready_to_scheduled( context, curr_node );
+      }
+
+      node = _Scheduler_strong_APA_Node_downcast( next_node );
+    }
+    /*
+     * To save the last node so that the caller SMP_* function
+     * can do the allocation
+     */
+      curr_node = &node->Base.Base;
+      highest_ready = curr_node;
+    }
+
+  return highest_ready;
 }
 
-static void _Scheduler_strong_APA_Extract_from_ready(
+/*
+ * Checks the lowest scheduled directly reachable task
+ */
+static inline Scheduler_Node *_Scheduler_strong_APA_Get_lowest_scheduled(
   Scheduler_Context *context,
-  Scheduler_Node    *the_thread
+  Scheduler_Node    *filter_base
 )
 {
-  Scheduler_strong_APA_Context *self =
-    _Scheduler_strong_APA_Get_self( context );
-  Scheduler_strong_APA_Node *node =
-    _Scheduler_strong_APA_Node_downcast( the_thread );
-
-  _Scheduler_priority_Ready_queue_extract(
-    &node->Base.Base.Node.Chain,
-    &node->Ready_queue,
-    &self->Bit_map
-  );
+  uint32_t	                cpu_max;
+  uint32_t	                cpu_index;
+  Scheduler_Node               *curr_node;
+  Scheduler_Node               *lowest_scheduled;
+  Priority_Control              max_priority_num;
+  Priority_Control              curr_priority;
+  Scheduler_strong_APA_Node    *filter_strong_node;
+  Scheduler_strong_APA_Context *self;
+
+  self = _Scheduler_strong_APA_Get_self( context );
+  lowest_scheduled = NULL; /* To remove compiler warning */
```

Comment:
do this assignment above? the comment isn't needed :)

Reply:
Done.